job/service.go

```diff
@@ -848,19 +855,30 @@ func (srv *Service) resolveDependency(ctx context.Context, projectSpec models.Pr
 	srv.notifyProgress(progressObserver, &models.ProgressJobDependencyResolutionFinished{})
 }
 
-func (srv *Service) resolveAndPersist(ctx context.Context, currentSpec models.JobSpec, projectSpec models.ProjectSpec, progressObserver progress.Observer) (interface{}, error) {
+func (srv *Service) resolve(ctx context.Context, currentSpec models.JobSpec, projectSpec models.ProjectSpec) ([]string, error) {
 	var err error
 	if currentSpec.Assets, err = srv.assetCompiler(currentSpec, srv.Now()); err != nil {
-		return currentSpec.Name, fmt.Errorf("%w: %s", errAssetCompilation, err.Error())
+		return nil, fmt.Errorf("%w: %s", errAssetCompilation, err.Error())
 	}
-	resolvedSpec, err := srv.dependencyResolver.Resolve(ctx, projectSpec, currentSpec, progressObserver)
+	jobSourceURNs, err := srv.dependencyResolver.ResolveInferredDependencies(ctx, projectSpec, currentSpec)
 	if err != nil {
-		return currentSpec.Name, fmt.Errorf("%s: %s: %w", errDependencyResolution, currentSpec.Name, err)
+		return nil, fmt.Errorf("%s: %s: %w", errDependencyResolution, currentSpec.Name, err)
 	}
-	if err := srv.dependencyResolver.Persist(ctx, resolvedSpec); err != nil {
-		return currentSpec.Name, fmt.Errorf("%s: %s: %w", errDependencyResolution, currentSpec.Name, err)
+	return jobSourceURNs, nil
+}
+
+func (srv *Service) persist(ctx context.Context, currentSpec models.JobSpec, projectID models.ProjectID, jobSourceURNs []string) error {
```

Comment:
before persisting, should also delete the existing sources of the job. try doing it with the transaction process.